README.md
# AI Platform Notebook Security Blueprint: Protecting PII Data

This repository provides an opinionated way to set up AI Platform Notebook in a secure way using Terraform.

**This is not an officially supported Google product**

## Reference Architecture
![Reference Architecture](diagrams/notebook_blueprint_arch.png)


The resources that this module will create are:

- One AI Platform Notebook per Notebook user
- Service Account for Notebooks
- an HSM key used for Customer Managed Encryption Keys (CMEK) in each Notebook
- Custom Role to restrict exporting data
- Google Cloud Storage bucket with bootstrap code for Notebooks
- Org Policies at the folder that the `trusted-data` project is in
  - `constraints/gcp.resourceLocations`
  - `constraints/iam.disableServiceAccountCreation`
  - `constraints/iam.disableServiceAccountKeyCreation`
  - `constraints/iam.automaticIamGrantsForDefaultServiceAccounts`
  - `constraints/compute.requireOsLogin`
  - `constraints/compute.restrictProtocolForwardingCreationForTypes`
  - `constraints/compute.restrictSharedVpcSubnetworks`

## Assumptions

*   You have your Project and network configuration available for where you want to deploy your trusted environment.
*   You have the appropriate IAM permissions to configure project resources (see [service account roles](#service-account)).
*   You have an IAM Group and list of identities that is allowed to access the trusted environment.
*   You are familiar with your organization's security best practices and policies.  Learn about Google Cloud security foundation best practices by reading the [security foundation blueprint](https://github.com/terraform-google-modules/terraform-example-foundation).

## Prerequisites

### Prepare your admin workstation
You can use Cloud Shell, a local machine or VM as your admin workstation

####
[**Tools for Cloud Shell as your Admin workstation**](#tools-for-cloud-shell-as-your-admin-workstation)

*   [Terraform >= 0.12.3](https://www.terraform.io/downloads.html)
*   [Terraform Provider for GCP][terraform-provider-gcp] plugin v3.51

####
[**Tools for a local workstation as your Admin workstation**](#tools-for-a-local-workstation-as-your-admin-workstation)

*   [Cloud SDK (gcloud CLI)](https://cloud.google.com/sdk/docs/quickstarts)
*   [Terraform >= 0.12.3](https://www.terraform.io/downloads.html)
*   [Terraform Provider for GCP][terraform-provider-gcp] plugin v3.51

####
**Installation instructions for Tools for your environment**


#####
[Install Cloud SDK](#install-cloud-sdk)


This is pre installed if you are using Cloud Shell

The Google Cloud SDK is used to interact with your GCP resources. [Installation instructions](https://cloud.google.com/sdk/downloads) for multiple platforms are available online.

#####
[Install Terraform](https://www.terraform.io/intro/getting-started/install.html)

Terraform is used to automate the manipulation of cloud infrastructure. Its [installation instructions](https://www.terraform.io/intro/getting-started/install.html) are also available online.
When configuring terraform for use with Google cloud create a service account as detailed in [Getting started with the google provider](https://www.terraform.io/docs/providers/google/guides/getting_started.html#adding-credentials)


#### **Authentication**

After installing the gcloud SDK run gcloud init to set up the gcloud cli. When executing choose the correct region and zone

'gcloud init'

Ensure you are using the correct project  . Replace my-project-name with the name of your project

Where the project name is my-project-name

`gcloud config set project my-project-name`


## Compatibility

This module is meant for use with Terraform 0.12. Learn how to [upgraded](https://www.terraform.io/upgrade-guides/0-12.html) to the required version.

## Usage

Basic usage of this module is as follows:

```hcl
module "notebooks_blueprint_security" {
  source  = "GoogleCloudPlatform/notebooks-blueprint-security/google"

  vpc_perimeter_regions           = ["US", "DE"]
  vpc_perimeter_policy_name       = "higher_trust_perimeter_policy"
  vpc_perimeter_ip_subnetworks    = ["NETWORK_CIDR"]  # allowed to access VPC-SC perimeters
  zone                            = "us-central1-a"
  resource_locations              = ["in:us-locations", "in:eu-locations"]
  notebook_key_name               = "trusted-data-key"
  dataset_id                      = "sample_ds_for_notebooks"
  notebook_name_prefix            = "trusted-sample"
  bootstrap_notebooks_bucket_name = "notebook_bootstrap"
  default_policy_id               = "12345678"  # likely org id
  project_trusted_analytics       = "trusted-analytics"
  project_trusted_data            = "trusted-data"
  project_trusted_kms             = "trusted-kms"
  trusted_private_network         = "projects/<shared-restricted-prj>/global/networks/<your_vpc>"
  trusted_private_subnet          = "projects/<shared-restricted-prj>/regions/<region>/subnetworks/<your_subnets_for_notebooks>"
  confidential_groups             = ["group:data-owners@example.com", "group:trusted-data-scientists@example.com"]
  trusted_scientists              = ["user:trusted1@example.com", "user:trusted2@example.com"]
}
```

1. Create a tfvars file with the required inputs (see [Inputs](#inputs) section below)
2. `terraform init ` to get the plugins
3. `terraform plan  -var-file="YOUR_FILE.tfvars"` to see the infrastructure plan.  Note: Replace `YOUR_file` with the name of your tfvars file from the first step
4. `terraform apply  -var-file="YOUR_FILE.tfvars"` to apply the infrastructure build. Note: Replace `YOUR_file` with the name of your tfvars file from the first step

5. Access your AI Platform Notebook
    * establish an [SSH tunnel](https://cloud.google.com/ai-platform/notebooks/docs/ssh-access) from your device to your AI Platform Notebook
    * in your browser, visit `http://localhost:8080` to access your AI Platform Notebook

Be sure to specify your `PROJECT_ID`, `DATASET`, and `TABLE` below, which should match your terraform.tfvars file.
```
%%bigquery
SELECT
  *
FROM `PROJECT_ID.DATASET.TABLE`
LIMIT 10
```

1. `terraform destroy  -var-file="YOUR_FILE.tfvars"` to destroy the built infrastructure. Note: Replace `YOUR_file` with the name of your tfvars file from the first step

### Adding identities to groups
1.  You may need to add service accounts the appropriate IAM high trust data scientist group.
```
# please change the values below to your specific values
gcloud identity groups memberships add --group-email grp-trusted-data-scientists@example.com --member-email = sa-p-notebook-compute@<proj>.iam.gserviceaccount.com
```

### Accessing Notebooks
Use ssh to access your notebook.  Notebooks have no external IP and users should not impersonate the Notebook service account.
Learn how to open an ssh tunnel to launch JuptyerLab, by reading the [SSH to access JupyterLab article](https://cloud.google.com/ai-platform/notebooks/docs/ssh-access).


Functional examples are included in the
[examples](./examples/) directory.


## Inputs

<!-- BEGINNING OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
## Inputs

| Name | Description | Type | Default | Required |
|------|-------------|------|---------|:--------:|
| bootstrap\_notebooks\_bucket\_name | Bucket name to create bootstrap scripts for notebooks. | `string` | `"notebook_bootstrap"` | no |
| confidential\_groups | The list of groups allowed to access PII data. | `list(string)` | n/a | yes |
| dataset\_id | BigQuery dataset ID with PII data that your scientists need to access from their Notebook. | `string` | n/a | yes |
| default\_policy\_id | The id of the default org policy. | `string` | n/a | yes |
| notebook\_key\_name | HSM key used to protect PII data in Notebooks. | `string` | `"trusted-data-key"` | no |
| notebook\_name\_prefix | Prefix for notebooks indicating in higher trusted environment. | `string` | `"trusted-sample"` | no |
| project\_trusted\_analytics | The trusted project for analytics activities and data scientists. | `string` | n/a | yes |
| project\_trusted\_data | The trusted project that has PII data for notebooks. | `string` | n/a | yes |
| project\_trusted\_kms | Top level trusted environment folder that will house the encryption keys. | `string` | n/a | yes |
| resource\_locations | The locations used in org policy to limit where resources can be provisioned. | `list(string)` | <pre>[<br>  "in:us-locations",<br>  "in:eu-locations"<br>]</pre> | no |
| trusted\_private\_network | Network with no external IP for Notebooks.  Should be a restricted private VPC. | `string` | n/a | yes |
| trusted\_private\_subnet | Subnet with no external IP for Notebooks.  Should be part of a restricted private network and have logs and private network enabled. | `string` | n/a | yes |
| trusted\_scientists | The list of trusted users. | `list(string)` | n/a | yes |
| vpc\_perimeter\_ip\_subnetworks | IP subnets for perimeters. | `list(string)` | n/a | yes |
| vpc\_perimeter\_policy\_name | Policy name for VPC service control perimeter. | `string` | `"higher_trust_perimeter_policy"` | no |
| vpc\_perimeter\_regions | 2 letter identifier for regions allowed for VPC access. A valid ISO 3166-1 alpha-2 code. | `list(string)` | n/a | yes |
| zone | The zone in which to create the secured notebook. Must match the region. | `string` | n/a | yes |

## Outputs

| Name | Description |
|------|-------------|
| access\_level\_name | access level name used in the perimeter policy |
| bkt\_notebooks\_name | name of bootstrap bucket |
| caip\_sa\_email | email of the SA used by CAIP; should not be a default SA |
| folder\_trusted | folder that holds all the trusted projects and constraints |
| notebook\_instances | list of notebooks created (vm names) |
| notebook\_key\_name | name of the key used in the notebooks. |
| notebook\_key\_ring\_name | name of keyring |
| perimeter\_name | vpc-sc perimeter name |
| script\_name | name of the post startup script installed |
| vpc\_perimeter\_resource\_protected | list of projects included in the VPC-Sc perimeter |

<!-- END OF PRE-COMMIT-TERRAFORM DOCS HOOK -->

## Requirements

These sections describe requirements for using this module.

### Software

The following dependencies must be available:

- [Terraform][terraform] v0.12
- [Terraform Provider for GCP][terraform-provider-gcp] plugin v3.51

### Service Account

A service account with the following roles must be used to provision
the resources of this module:

Organization Level
- Access Context Manager Policy Admin: `roles/accesscontextmanager.policyAdmin`
- Organization Policy Admin: `roles/orgpolicy.policyAdmin`
- Security Admin: `roles/iam.securityAdmin`
- Service Usage Consumer: `roles/serviceusage.serviceUsageConsumer`

Restricted Shared VPC Project (created in blueprint foundation)
- Network Admin: `compute.networkAdmin`

Analytics Project
- Service Account Creator: `roles/iam.serviceAccountCreator`
- Cloud KMS Admin: `roles/cloudkms.admin`
- Compute Instance Admin: `roles/compute.admin`
- BigQuery Job User: `roles/bigquery.jobUser`
- BigQuery User: `roles/bigquery.user`
- Notebooks Runner: `roles/notebooks.runner`
- Service Account User: `roles/iam.serviceAccountUser`
- Service Usage Admin: `roles/serviceusage.serviceUsageAdmin`

Data Project
- BigQuery Job User: `roles/bigquery.jobUser`
- BigQuery User: `roles/bigquery.user`
- Role Administrator: `roles/iam.roleAdmin`
- Storage Admin: `roles/storage.admin`

KMS Project
- Cloud KMS Admin: `roles/cloudkms.admin`

The [Project Factory module][project-factory-module] and the
[IAM module][iam-module] may be used in combination to provision a
service account with the necessary roles applied.

### Enable APIs
In order to operate with the Service Account you must activate the following APIs on the project where analytics and Notebooks reside:

- Access Context Manager API: `accesscontextmanager.googleapis.com`
- BigQuery API: `bigquery.googleapis.com`
- Compute Engine API: `compute.googleapis.com`
- Identity and Access Management (IAM) API: `iam.googleapis.com`
- Key Management Service (KMS) API: `cloudkms.googleapis.com`
- Notebooks (AI Platform) API: `notebooks.googleapis.com`
- Google Cloud Storage API: `storage.googleapis.com`
- Resource Manager API: `cloudresourcemanager.googleapis.com`
- IAM Service Account Credentials API: `iamcredentials.googleapis.com`

In order to operate with the Service Account you must activate the following APIs on the project where your KMS/HSM keys reside:

- Google Cloud Storage API: `storage.googleapis.com`
- Key Management Service (KMS) API: `cloudkms.googleapis.com`


### Resource Hierarchy
Within your Org's prod environment, create a folder to hold your trusted projects and centrally managed your policies for Notebooks that use PII data.
Note: the `fldr-prod` is created by the foundation blueprint.  Create folders by using the [project factory](https://github.com/terraform-google-modules/terraform-google-project-factory)

```
fldr-prod
└── fldr-trusted
    ├── trusted-data
    ├── trusted-analytics
    └── trusted-kms
```


## Contributing

Refer to the [contribution guidelines](./CONTRIBUTING.md) for
information on contributing to this module.

[iam-module]: https://registry.terraform.io/modules/terraform-google-modules/iam/google
[project-factory-module]: https://registry.terraform.io/modules/terraform-google-modules/project-factory/google
[terraform-provider-gcp]: https://www.terraform.io/docs/providers/google/index.html
[terraform]: https://www.terraform.io/downloads.html


CHANGELOG.md
# Changelog

### [1.0.1](https://www.github.com/GoogleCloudPlatform/notebooks-blueprint-security/compare/v1.0.0...v1.0.1) (2021-09-12)


### Bug Fixes

* clarify output descriptions ([#63](https://www.github.com/GoogleCloudPlatform/notebooks-blueprint-security/issues/63)) ([63e9d7e](https://www.github.com/GoogleCloudPlatform/notebooks-blueprint-security/commit/63e9d7e240477a99fd778126303a152d48e4e2f4))
* cleanup docs ([#62](https://www.github.com/GoogleCloudPlatform/notebooks-blueprint-security/issues/62)) ([5e5cbd8](https://www.github.com/GoogleCloudPlatform/notebooks-blueprint-security/commit/5e5cbd86a567b6c39ae1ee7f73e7640aa079dddb))
* lint issues ([#61](https://www.github.com/GoogleCloudPlatform/notebooks-blueprint-security/issues/61)) ([4bdda7d](https://www.github.com/GoogleCloudPlatform/notebooks-blueprint-security/commit/4bdda7dd3d1b1361d62fda00d97542eb4e58e335))
* multiple users ([#57](https://www.github.com/GoogleCloudPlatform/notebooks-blueprint-security/issues/57)) ([3e5863c](https://www.github.com/GoogleCloudPlatform/notebooks-blueprint-security/commit/3e5863c853076ec6b418ce0bd3af0dc82f3836fe))
* support identity characters ([#55](https://www.github.com/GoogleCloudPlatform/notebooks-blueprint-security/issues/55)) ([9d6ca28](https://www.github.com/GoogleCloudPlatform/notebooks-blueprint-security/commit/9d6ca281e5784a60cb56fd792a5ceebb388a2bc8)), closes [#53](https://www.github.com/GoogleCloudPlatform/notebooks-blueprint-security/issues/53)


test/test.sh
#!/bin/bash
#
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# `source test/setup_variables.sh && ./test/test.sh`


IMPERSONATION_SA=""
if [[ -z ${TERRAFORM_SA} ]]; then
    IMPERSONATION_SA=${SA_NAME}@${DEPLOYMENT_PROJECT}.iam.gserviceaccount.com
else
    IMPERSONATION_SA=${TERRAFORM_SA}
fi

# Impersonnate with a token
gcloud config set auth/impersonate_service_account ${IMPERSONATION_SA}
export GOOGLE_OAUTH_ACCESS_TOKEN=$(gcloud auth print-access-token)

# Test against CIS benchmark
PROJECT_LIST=(
  ${PRJ_HIGH_TRUST_ANALYTICS}
  ${PRJ_HIGH_TRUST_DATA}
  ${PRJ_HIGH_TRUST_DATA_ETL}
  ${PRJ_HIGH_TRUST_KMS}
)

# Run CIS on all projects created
for PRJ in "${PROJECT_LIST[@]}"
do
  inspec exec https://github.com/GoogleCloudPlatform/inspec-gcp-cis-benchmark.git -t gcp:// --input gcp_project_id=${PRJ} --reporter cli json:${PRJ}_scan.json
done

# Run all kitchen-terraform tests
kitchen destroy

kitchen create

kitchen converge

kitchen verify

kitchen destroy

gcloud config unset auth/impersonate_service_account

gcloud config unset auth/impersonate_service_account


test/setup_variables.sh
#
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

#===============================================================
# The following MUST be set to your project specific 
# resource names and environment
#===============================================================

# Provide your project name that holds terraform service account credentials.
# This project should be in the bootstrap folder's seed project
# This where the terraform state will be
export DEPLOYMENT_PROJECT_ID=""

# Top level trusted folder
# "fldr-trusted"
# "fldr-production"
# " elo-fldr-prod-trusted2"
export PARENT_FOLDER=""

# folder that has all the trusted projets (e.g folder/123456890)
export TRUSTED_FOLDER_ID=""

# Provide your billing account
# gcloud beta billing accounts list
# ABCDEF-234567-HIJKLM
export BILLING_ACCOUNT=""  

# Your organization ID
# `glcoud organizations list`
export ORGANIZATION_ID=""

# Existing access policy ID number
# 987654321
export POLICY_NAME=""

# Your project names defined in the terraform/terraform.template.tfvars file
export PRJ_HIGH_TRUST_ANALYTICS=""
export PRJ_HIGH_TRUST_DATA=""
export PRJ_HIGH_TRUST_DATA_ETL=""
export PRJ_HIGH_TRUST_KMS=""

# Provide your privileged service account used by terraform.
# This will likely be the same one used from the security foundational blueprint
# otherwise, comment it out, so that one can be automatically generated
export TERRAFORM_SA=""
export CAIP_COMPUTE_SA=""

# Existing access policy ID number
# 99998888
# Note: only provide the numeric identifier from accessPolicies/<policyName>"
# gcloud access-context-manager policies list --organization ${ORGANIZATION_ID} --format json | jq .[].name | sed 's/"//g' | awk '{split($0,a,"/"); print a[2]}'
export POLICY_NAME=""

# Kitchen-terraform variables
export TF_VAR_organizations=${ORGANIZATION_ID}
export TF_VAR_org=organziations/${ORGANIZATION_ID}
export TF_VAR_default_policy_name=${POLICY_NAME}
export TF_VAR_terraform_sa_email=${TERRAFORM_SA}
export TF_VAR_caip_compute_sa_email=${CAIP_COMPUTE_SA}
export TF_VAR_billing_account=${BILLING_ACCOUNT}
export TF_VAR_trusted_folder_id=${TRUSTED_FOLDER_ID}
export TF_CLI_ARGS_apply="-var-file=$(pwd)/terraform/terraform.tfvars"
export TF_CLI_ARGS_destroy="-var-file=$(pwd)/terraform/terraform.tfvars"

#===============================================================
# Output for visual inspection
#===============================================================
echo ""
echo "Environment setup with values:"
echo "==================================================="
echo "Billing account: ${BILLING_ACCOUNT:?}"
echo "Deployment project ID: ${DEPLOYMENT_PROJECT_ID:?}"
echo "Policy name: ${POLICY_NAME:?}"
echo "Parent folder: ${PARENT_FOLDER:?}"
echo "Trusted folder ID: ${TRUSTED_FOLDER_ID:?}"
echo "Organization: ${ORGANIZATION_ID:?}"
echo "Policy name: ${POLICY_NAME:?}"
echo "Project high trust analytics: ${PRJ_HIGH_TRUST_ANALYTICS:?}"
echo "Project high trust data: ${PRJ_HIGH_TRUST_DATA:?}"
echo "Project high trust data ETL: ${PRJ_HIGH_TRUST_DATA_ETL:?}"
echo "Project high trust kms: ${PRJ_HIGH_TRUST_KMS:?}"
echo "Terraform service account: ${TERRAFORM_SA:?}"
echo "CAIP compute Service Account: ${CAIP_COMPUTE_SA:?}"
echo "==================================================="
echo ""
echo "Please verify these environment variables are correct for your GCP Resources"
echo ""


test/setup/make_source.sh
#!/usr/bin/env bash

# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

echo "#!/usr/bin/env bash" > ../source.sh

example_zone=$(terraform output example_zone)
echo "export TF_VAR_zone='${example_zone}'" >> ../source.sh

example_vpc=$(terraform output example_vpc)
example_vpc_subnet=$(terraform output example_vpc_subnet)
echo "export TF_VAR_trusted_private_network='${example_vpc}'" >> ../source.sh
echo "export TF_VAR_trusted_private_subnet='${example_vpc_subnet}'" >> ../source.sh

example_dataset_id=$(terraform output dataset_id)
echo "export TF_VAR_dataset_id='${example_dataset_id}'" >> ../source.sh

# TODO use token when https://github.com/inspec/inspec-gcp/issues/282 is added
sa_json=$(terraform output sa_key)
# shellcheck disable=SC2086
echo "export SERVICE_ACCOUNT_JSON='$(echo $sa_json | base64 --decode)'" >> ../source.sh


terraform/setup_varibles.sh
#
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

#===============================================================
# The following MUST be set to your project specific 
# resource names and environment
#===============================================================

# Provide your already created project id that holds terraform service account credentials.
# This project should be in the bootstrap folder's seed project
# ex: prj-aaaa-p-seed
export DEPLOYMENT_PROJECT_ID=""

# Top level folder that will hold all the projects with higher trust data
# ex: fldr-production
export PARENT_FOLDER=""

# Your existing organization ID
# ex: 111112222
# Note: use `glcoud organizations list`
export ORGANIZATION_ID=""

# Existing access policy ID number
# ex: 99998888
# Note: only provide the numeric identifier from accessPolicies/<policyName>"
# gcloud access-context-manager policies list --organization <orgID> --format json | jq .[].name | sed 's/"//g' | awk '{split($0,a,"/"); print a[2]}'
export POLICY_NAME=""

# Provide your billing account
# gcloud beta billing accounts list
# ABCDEF-123456-A1B2C3
export BILLING_ACCOUNT=""  

# Provide your privileged service account used by terraform.
# This will likely be the same one used from the security foundational blueprint
# otherwise, comment it out, so that one can be automatically generated
# ex: org-terraform@<projectName>.iam.gserviceaccount.com
# Note: use the below command to pick the correct terraform, if using the foundational blueprint
# `gcloud iam service-accounts list | grep terraform`
export TERRAFORM_SA=""

#===============================================================
# Output for visual inspection
#===============================================================
echo ""
echo "Environment setup with values:"
echo "==================================================="
echo "Deployment project ID: ${DEPLOYMENT_PROJECT_ID:?}"
echo "Parent Folder: ${PARENT_FOLDER:?}"
echo "Organization: ${ORGANIZATION_ID:?}"
echo "Access Context Policy: ${POLICY_NAME:?}"
echo "Billing Account: ${BILLING_ACCOUNT:?}"
echo "Terraform Service Account: ${TERRAFORM_SA:?}"
echo "==================================================="
echo ""
echo "Please verify these environment variables are correct for your GCP Resources"
echo ""


bootstrap_files/post_startup_script.sh
#!/bin/bash
# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


INSTANCE_METADATA_ATTRIBUTES="http://metadata.google.internal/computeMetadata/v1/instance/attributes/"
KO_DOWNLOADS=$(curl "${INSTANCE_METADATA_ATTRIBUTES}/notebook-disable-downloads" -H "Metadata-Flavor: Google")
KO_NBCONVERT=$(curl "${INSTANCE_METADATA_ATTRIBUTES}/notebook-disable-nbconvert" -H "Metadata-Flavor: Google")

export USER="jupyter"

function err() {
  echo "[$(date +'%Y-%m-%dT%H:%M:%S%z')]: $*" >&2
  exit 1
}

function disable_downloads() {
  echo "---------------------------------"
  echo "Creating ForbidFilesHandler class"
  echo "---------------------------------"
  cat <<END >/home/${USER}/.jupyter/handlers.py
from tornado import web
from notebook.base.handlers import IPythonHandler

class ForbidFilesHandler(IPythonHandler):
  @web.authenticated
  def head(self, path):
    self.log.info("HEAD: File download forbidden.")
    raise web.HTTPError(403)

  @web.authenticated
  def get(self, path, include_body=True):
    self.log.info("GET: File download forbidden.")
    raise web.HTTPError(403)

END

  echo "--------------------------------"
  echo "Override the files_handler_class"
  echo "--------------------------------"
  cat <<END >>/home/${USER}/.jupyter/jupyter_notebook_config.py
import os, sys
sys.path.append('/home/${USER}/.jupyter/')
import handlers
c.ContentsManager.files_handler_class = 'handlers.ForbidFilesHandler'
c.ContentsManager.files_handler_params = {}

# prevent export/printing of calculated values that likely have PII
c.TemplateExporter.exclude_input_prompt = True
c.TemplateExporter.exclude_output = True

END

}

function disable_nbconvert() {
  # nbconvert configuration can be altered to prevent output from being stored locally/exported
  pip uninstall nbconvert
}

function main() {

  echo "Checking if must disable downloads..."
  if [[ "${KO_DOWNLOADS}" == 'true' ]]; then
    echo "Calling disable_downloads..."
    disable_downloads || echo 'Error when disabling downloads.'
  fi

  echo "Checking if must uninstall nbconvert..."
  if [[ "${KO_NBCONVERT}" == 'true' ]]; then
    echo "Calling disable_nbconvert..."
    disable_nbconvert || echo 'Error when uninstalling nbconvert.'
  fi

  echo "Check if must restart jupyter.service..."
  if [ "${KO_DOWNLOADS}" == 'true' ] || [ "${KO_NBCONVERT}" == 'true' ]; then
    echo "Try to restart jupyter.service..."
    systemctl restart jupyter.service || echo 'Error restarting jupyter.service.'
  fi

}

main



policies/validate.sh
#!/bin/bash

# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Provide your privileged service account used by terraform.
# This will likely be the same one used from the security foundational blueprint
# otherwise, comment it out, so that one can be automatically generated
export TERRAFORM_SA=""  # e.g. <name>@<projectID>.iam.gserviceaccount.com

export ORGANIZATION_ID="" # 12345678

# Existing access policy ID number
# e.g. 99998888
# Note: only provide the numeric identifier from accessPolicies/<policyName>"
# gcloud access-context-manager policies list --organization <orgID> --format json | jq .[].name | sed 's/"//g' | awk '{split($0,a,"/"); print a[2]}'
export POLICY_NAME=""

# Provide your billing account
# gcloud beta billing accounts list
# ABCDEF-123456-A1B2C3
export BILLING_ACCOUNT=""

export TRUSTED_FOLDER_ID="" # 12345678

# folder that has all the trusted projects (e.g folder/123456890)
export TRUSTED_FOLDER_RESOURCE_NAME="organizations/${ORGANIZATION_ID}/folders/${TRUSTED_FOLDER_ID}"

# Your project names defined in the terraform/terraform.template.tfvars file
# replace <ID> with a real value (e.g. 12345678)
export PRJ_HIGH_TRUST_ANALYTICS="${TRUSTED_FOLDER_RESOURCE_NAME}/projects/<ID>"
export PRJ_HIGH_TRUST_KMS="${TRUSTED_FOLDER_RESOURCE_NAME}/projects/<ID>"
export PRJ_HIGH_TRUST_DATA="${TRUSTED_FOLDER_RESOURCE_NAME}/projects/<ID>"
export PRJ_HIGH_TRUST_DATA_ETL="${TRUSTED_FOLDER_RESOURCE_NAME}/projects/<ID>"
export ZONE="" # e.g. us-central1-a
export REGIONS="" # e.g us;  these should be separate by spaces such as "us eu"
export DOMAIN=""  # your organization's domain

# a list of regex's of instances to test
export VM_INSTANCES="//compute.googleapis.com/projects/${PRJ_HIGH_TRUST_ANALYTICS}/.*"

# list of networks where trusted environment can use.  values are separated by spaces (e.g. <...>/network1 <...>/network2)
# e.g https://www.googleapis.com/compute/v1/projects/<projectid>/global/networks/<networkName>
export NETWORKS=""

# list of access level policy names.  values are separated by a space.
export VPC_SC_ACCESS_LEVELS_NAME=""

# list of projects within the vpc-sc.  values are separated by a space.
# e.g.  projects/11111111 projects/2222222 projects/333333
export VPC_SC_PROJECTS=""

#===============================================================
# Output for visual inspection
#===============================================================
echo ""
echo "Environment setup with values:"
echo "==================================================="
echo "Terraform service account: ${TERRAFORM_SA:?}"
echo "Organization: ${ORGANIZATION_ID:?}"
echo "Billing account: ${BILLING_ACCOUNT:?}"
echo "Policy name: ${POLICY_NAME:?}"
echo "Trusted folder ID: ${TRUSTED_FOLDER_ID:?}"
echo "Deployment project ID: ${TRUSTED_FOLDER_RESOURCE_NAME:?}"
echo "Project high trust analytics: ${PRJ_HIGH_TRUST_ANALYTICS:?}"
echo "Project high trust data: ${PRJ_HIGH_TRUST_DATA:?}"
echo "Project high trust data ETL: ${PRJ_HIGH_TRUST_DATA_ETL:?}"
echo "Project high trust kms: ${PRJ_HIGH_TRUST_KMS:?}"
echo "Zone: ${ZONE:?}"
echo "Regions: ${REGIONS:?}"
echo "Domain: ${DOMAIN:?}"
echo "VM instances: ${VM_INSTANCES:?}"
echo "Networks: ${NETWORKS:?}"
echo "VPC-SC access levels: ${VPC_SC_ACCESS_LEVELS_NAME:?}"
echo "VPC-SC projects: ${VPC_SC_PROJECTS:?}"
echo "==================================================="
echo ""
echo "Please verify these environment variables are correct for your GCP Resources"
echo ""

echo "Creating Kptfile..."
kpt cfg create-setter constraint/ project_owners --type array --field spec.match.members

kpt cfg create-setter constraint/ target_fldr_trusted --type array --field spec.match.target
kpt cfg create-setter constraint/ target_prj_analytics --type array --field spec.match.target
kpt cfg create-setter constraint/ target_prj_kms --type array --field spec.match.target
kpt cfg create-setter constraint/ target_prj_data --type array --field spec.match.target
kpt cfg create-setter constraint/ target_prj_data_etl --type array --field spec.match.target

kpt cfg create-setter constraint/ notebook_zones --type array --field spec.parameters.zones
kpt cfg create-setter constraint/ locations --type array --field spec.parameters.locations

kpt cfg create-setter constraint/ domain ${DOMAIN}

kpt cfg create-setter constraint/ vm_instances --type array --field spec.parameters.instances
kpt cfg create-setter constraint/ allowed_networks --type array --field spec.parameters.allowed

kpt cfg create-setter constraint/ vpc_sc_regions --type array --field spec.parameters.regions
kpt cfg create-setter constraint/ vpc_sc_access_levels --type array --field spec.parameters.required_access_levels
kpt cfg create-setter constraint/ vpc_sc_project_ids --type array --field spec.parameters.required_projects

echo "...configuring setters and updated yaml files"
kpt cfg set constraint/ project_owners ${TERRAFORM_SA}

kpt cfg set constraint/ target_fldr_trusted ${TRUSTED_FOLDER_RESOURCE_NAME}
kpt cfg set constraint/ target_prj_analytics ${PRJ_HIGH_TRUST_ANALYTICS}
kpt cfg set constraint/ target_prj_kms ${PRJ_HIGH_TRUST_KMS}
kpt cfg set constraint/ target_prj_data ${PRJ_HIGH_TRUST_DATA}
kpt cfg set constraint/ target_prj_data_etl ${PRJ_HIGH_TRUST_DATA_ETL}

kpt cfg set constraint/ notebook_zones ${ZONE}
kpt cfg set constraint/ locations ${REGIONS}

kpt cfg set constraint/ domain ${DOMAIN}

kpt cfg set constraint/ vm_instances "${VM_INSTANCES}"
kpt cfg set constraint/ allowed_networks ${NETWORKS}

kpt cfg set constraint/ vpc_sc_regions ${REGIONS}
kpt cfg set constraint/ vpc_sc_access_levels ${VPC_SC_ACCESS_LEVELS_NAME}
kpt cfg set constraint/ vpc_sc_project_ids ${VPC_SC_PROJECTS}

pushd ../../terraform || exit
# source setup_variables.sh
# impersonate with a token (use cloud identity to set the default time to 1 hr)
# this uses Oauth2 so that a SA key isn't needed
gcloud config set auth/impersonate_service_account ${TERRAFORM_SA}
GOOGLE_OAUTH_ACCESS_TOKEN=$(gcloud auth print-access-token)
export GOOGLE_OAUTH_ACCESS_TOKEN

terraform init
terraform plan \
  -out tfplan.tfplan \
  -var-file="terraform.template.tfvars" \
  -var "org=organizations/${ORGANIZATION_ID}" \
  -var "default_policy_name=${POLICY_NAME}" \
  -var "terraform_sa_email=${TERRAFORM_SA}" \
  -var "billing_account=${BILLING_ACCOUNT}"

popd || exit

echo "validator skipped until v0.13 is supported"
# TODO uncomment once validator support terraform v0.13
#terraform-validator-linux-amd64 validate ../../terraform/tfplan.tfplan --policy-path=./tmp

gcloud config unset auth/impersonate_service_account
